### DISCUSSION TITLE
Single Quotes in Shell Script Compilation Error

### DISCUSSION THREAD

**Discussion by gouinK** [nextflow-io/nextflow#2539] [Q&A] [answered]
Greetings,

I am using the Shell block for one of my processes because I use a combination of Bash and Nextflow variables in it. 

One of the commands in the script is as follows:

echo '!"#$%&'\''()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\]^_`abcdefghijklmnopqrstuvwxyz{|}~' | fold -w 1 > QS_matches.txt

This reproducibly causes a compilation error (see below) for the process, and it points to the 3 single quotes that begin the shell block as the issue, where line 186 below refers to the start of the shell block:

N E X T F L O W  ~  version 21.10.6
Launching `/home/main.nf` [silly_liskov] - revision: 7e000672a8
Script compilation error
- file : /home/main.nf
- cause: Unexpected character: '\'' @ line 186, column 7.
       '''
         ^

1 error

Running the "echo" command outside of Nextflow works fine, so I suspect it has something to do with too many single quotes being used in the command.

Any thoughts on what I might be doing wrong here?

Thanks!

**Comment by manuelesimi** [comment, marked as answer]
Hi,

See if the following helps (it works for me):

```
process Test {

    output:
    path("QS_matches.txt")

    shell:
    '''
    echo '!"#$%&'"'""'""'"'()*+,-./0123456789:;<=>?@abcdefghijklmnopqrstuvwxyz[]^_`abcdefghijklmnopqrstuvwxyz{|}~' | fold -w 1 > QS_matches.txt
    '''
}
```

This is the content of the output file:
```
!
"
#
$
%
&
'
'
'
(
)
*
+
,
-
.
/
0
1
2
3
4
5
6
7
8
9
:
;
<
=
>
?
@
a
b
c
d
e
f
g
h
i
j
k
l
m
n
o
p
q
r
s
t
u
v
w
x
y
z
[
]
^
_
`
a
b
c
d
e
f
g
h
i
j
k
l
m
n
o
p
q
r
s
t
u
v
w
x
y
z
{
|
}
~
```
I exploited the capability offered by Bash to glue both kinds of quotations (single and double) and then used `"'"` to quote  each single quote inside your string.
